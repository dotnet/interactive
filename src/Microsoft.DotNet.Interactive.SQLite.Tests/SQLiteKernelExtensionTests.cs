// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using Assent;
using Microsoft.DotNet.Interactive.CSharp;
using Microsoft.DotNet.Interactive.Events;
using Microsoft.DotNet.Interactive.Formatting;
using Microsoft.DotNet.Interactive.Formatting.TabularData;
using Xunit;

namespace Microsoft.DotNet.Interactive.SQLite.Tests;

public class SQLiteKernelExtensionTests : IDisposable
{
    private readonly Configuration _configuration;

    public SQLiteKernelExtensionTests()
    {
        _configuration = new Configuration()
            .SetInteractive(Debugger.IsAttached)
            .UsingExtension("json");
    }

    [Fact]
    public async Task can_generate_tabular_json_from_database_table_result()
    {
        using var kernel = new CompositeKernel
        {
            new CSharpKernel().UseNugetDirective(),
            new KeyValueStoreKernel()
        };

        SQLiteKernelConnector.AddSQLiteKernelConnectorTo(kernel);

        using var _ = SQLiteConnectionTests.CreateInMemorySQLiteDb(out var connectionString);

        await kernel.SubmitCodeAsync(
            $"#!connect sqlite --kernel-name mydb \"{connectionString}\"");

        var result = await kernel.SubmitCodeAsync(@"
#!sql-mydb
SELECT * FROM fruit
");

        var tabularData =  result.Events.OfType<DisplayedValueProduced>().Single()
            .Value;

           
        this.Assent(tabularData.ToDisplayString(TabularDataResourceFormatter .MimeType), _configuration);
    }

    [Fact]
    public async Task can_handle_duplicate_columns_in_query_results()
    {
        using var kernel = new CompositeKernel
        {
            new CSharpKernel().UseNugetDirective(),
            new KeyValueStoreKernel()
        };

        SQLiteKernelConnector.AddSQLiteKernelConnectorTo(kernel);

        using var _ = SQLiteConnectionTests.CreateInMemorySQLiteDb(out var connectionString);

        await kernel.SubmitCodeAsync(
            $"#!connect sqlite --kernel-name mydb \"{connectionString}\"");

        var result = await kernel.SubmitCodeAsync(@"
#!sql-mydb
SELECT 1 AS Apples, 2 AS Bananas, 3 AS Apples, 4 AS BANANAS, 5 AS Apples, 6 AS BaNaNaS
");

        var tabularData = result.Events.OfType<DisplayedValueProduced>().Single()
            .Value;


        this.Assent(tabularData.ToDisplayString(TabularDataResourceFormatter.MimeType), _configuration);
    }

    public void Dispose()
    {
        Formatter.ResetToDefault();
        DataExplorer.ResetToDefault();
    }
}