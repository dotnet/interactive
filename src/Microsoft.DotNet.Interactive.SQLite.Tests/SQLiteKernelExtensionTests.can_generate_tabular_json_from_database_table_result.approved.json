<table><thead><tr><td><span>id</span></td><td><span>name</span></td><td><span>color</span></td><td><span>deliciousness</span></td></tr></thead><tbody><tr><td><div class="dni-plaintext">1</div></td><td>apple</td><td>green</td><td><div class="dni-plaintext">10</div></td></tr><tr><td><div class="dni-plaintext">2</div></td><td>banana</td><td>red</td><td><div class="dni-plaintext">11</div></td></tr><tr><td><div class="dni-plaintext">3</div></td><td>cherry</td><td>red</td><td><div class="dni-plaintext">9000</div></td></tr></tbody></table>